networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/rpc/core/Eth2OutgoingRequestHandler.java

```diff
@@ -57,56 +58,70 @@ public Eth2OutgoingRequestHandler(
 
     responseProcessor =
         new AsyncResponseProcessor<>(asyncRunner, responseStream, this::onAsyncProcessorError);
-    responseHandler = new ResponseRpcDecoder<>(responseProcessor::processResponse, this.method);
+    responseDecoder = method.createResponseDecoder();
   }
 
-  @Override
-  public void onActivation(final RpcStream rpcStream) {
-    this.rpcStream = rpcStream;
+  public void handleInitialPayloadSent(final RpcStream stream) {
+    this.rpcStream = stream;
+
+    if (method.shouldReceiveResponse()) {
+      // Close the write side of the stream
+      stream.closeWriteStream().reportExceptions();
+      // Start timer for first bytes
+      ensureFirstBytesArriveWithinTimeLimit(stream);
+    } else {
+      // If we're not expecting any response, complete the request
+      completeRequest(stream);
+    }
   }
 
   @Override
-  public void onData(final NodeId nodeId, final RpcStream rpcStream, final ByteBuf bytes) {
+  public void processInput(
+      final NodeId nodeId, final RpcStream rpcStream, final InputStream input) {
     try {
-      if (hasReceivedInitialBytes.compareAndSet(false, true)) {
-        // Setup initial chunk timeout
-        ensureNextResponseArrivesInTime(rpcStream, currentChunkCount.get(), currentChunkCount);
-      }
-      LOG.trace("Requester received {} bytes.", bytes.capacity());
-      responseHandler.onDataReceived(bytes);
-
-      final int previousResponseCount = currentChunkCount.get();
-      currentChunkCount.set(responseProcessor.getResponseCount());
-      if (currentChunkCount.get() >= maximumResponseChunks) {
-        completeRequest(rpcStream);
-      } else if (currentChunkCount.get() > previousResponseCount) {
-        ensureNextResponseArrivesInTime(rpcStream, currentChunkCount.get(), currentChunkCount);
+      this.rpcStream = rpcStream;
+
+      Optional<TResponse> maybeResponse =
+          responseDecoder.decodeNextResponse(input, this::onFirstByteReceived);
```

Comment:
In terms of being able to re-use the one input stream instead of discarding the wrapper each time - I wonder if `responseDecoder` rather than being used directly could be changed to something like `responseDecoder.forInputStream(input)` and that returns an object that provides the `decodeNextResponse` methods - then it'd be able to hold onto the one snappy wrapper stream.  I think you'd have to do the same trick with `RpcEncoding` having a `decoderFor(InputStream)` kind of method too and maybe again at the actual Compressor level.

Unfortunate amount of boiler plate code there but `SnappyFramedInputStream` has a `frameHeader` buffer which worries me that would could potentially lose bytes.